### PR TITLE
[SPARK-19825][R][ML] spark.ml R API for FPGrowth

### PR DESCRIPTION
**Pull request by zero323**
## What changes were proposed in this pull request?

Adds SparkR API for FPGrowth: [SPARK-19825](https://issues.apache.org/jira/browse/SPARK-19825):

- `spark.fpGrowth` -model training.
- `freqItemsets` and `associationRules` methods with new corresponding generics.
- Scala helper: `org.apache.spark.ml.r. FPGrowthWrapper`
- unit tests.

## How was this patch tested?

Feature specific unit tests.

### CODE REVIEW

File: R/pkg/R/generics.R
Comment (felixcheung) [inline R/pkg/R/generics.R]: we seems to follow the pattern `spark.something` - see LDA. do you think it makes sense here too?

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: should it have `numPartitions`?

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: instead of `features` it should take a formula?

Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: we generally avoid allow setting `predictionCol` too

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: you may want to `as.numeric` on `minSupport`, `minConfidence` in case someone is passing in an integer and `callJStatic` would fail to match the wrapper method

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: It might make sense to do this in a single line:
```
dataFrame(callJMethod(object@jobj, "freqItemsets")
```

might be more readable that way. ditto with Association Rules below

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: could you use the long form name (eg. look at LDA) and drop the word "Model" which we avoid using

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: instead of duplicating `createDataFrame`, set `itemsets <- createDataFrame(data.frame(features = c("a,b", "a,b,c", "c,d")))`

btw, do we have real data to use instead?

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: collapse this to `head(freqItemsets(model))`

File: R/pkg/inst/tests/testthat/test_mllib_fpm.R
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_mllib_fpm.R]: don't repeat `freqItemsets(model)` - use `itemsets` from above

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: do we need these?

**Review by felixcheung — CHANGES_REQUESTED** [review]
thanks, I think a couple of big changes are in order. Let me know when it is ready for review again

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (zero323) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: We can skip `import org.json4s._` if won't do any parsing, but import org.json4s.jackson.JsonMethods._` provide both `render` and `compact` which are used to create JSON metadata.

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: Do you mean `spark.FPGrowth`? I can but as far as I can tell all classes use `Model` suffix ([`GeneralizedLinearRegressionModel`](https://github.com/apache/spark/blob/89cd3845b6edb165236a6498dcade033975ee276/R/pkg/R/mllib_regression.R#L33), [`GaussianMixtureModel`](https://github.com/apache/spark/blob/89cd3845b6edb165236a6498dcade033975ee276/R/pkg/R/mllib_clustering.R#L32) [`LDAModel`](https://github.com/apache/spark/blob/89cd3845b6edb165236a6498dcade033975ee276/R/pkg/R/mllib_clustering.R#L46) and so on) and none is using `spark` prefix.

Or do you mean `representation` instead of `slots`? I believe that representation is no longer recommended.

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: Yes, we do. Adjusted.

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: I mean this 
https://github.com/apache/spark/blob/master/R/pkg/R/mllib_clustering.R#L467
https://github.com/apache/spark/blob/master/R/pkg/R/mllib_clustering.R#L316
which may or may not include the word model

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: about here-  thought?

**Review by felixcheung — COMMENTED** [review]
https://github.com/apache/spark/pull/17170/files/6c0aea9ffc66bf525f61efca14c8630dbb940d52#r104594539

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: Done.

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: To be honest I am not sure. If you think that setting `predictionCol` should  be disabled I am fine with that but I don't see how formulas could be useful here. `FPGrowth` doesn't really conform to the conventions used in other ML algorithms. It doesn't use vectors and fixed size buckets are unlikely to happen.

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: Done.

**Comment by zero323** [comment]
@felixcheung I think i addressed all the issues excluding `inputCol` and `predictionCol`. In general:

- Using `formula`as an input doesn't make sense in my opinion. 
- Personally I would allow users to set column names. Both `features` and `prediction` are a bit vague in the context of the algorithm.

**Comment by felixcheung** [comment]
Sure - it's a single columns called `features` so I'm fine with it as a parameter.
I'm not sure about `inputCol` though, it's a different nomenclature in R that is different from the rest of mllib `features`.

**Comment by zero323** [comment]
I think that [ALS sets a precedence for using `somethingCol`](https://github.com/apache/spark/blob/master/R/pkg/R/mllib_recommendation.R#L86) but I don't like 'features" part here. Maybe `basketsCol`, what you think?

**Comment by felixcheung** [comment]
To clarify with your example, with ALS we have userCol, ratingCol - these matches the API names in spark.ml, and I think we need to do the same here.

What don't you like about [`featuresCol` which is also in the Scala API](https://github.com/apache/spark/blob/0fe8020f3aaf61c9992b6bcc5dba7ae8f751bab7/mllib/src/main/scala/org/apache/spark/ml/fpm/FPGrowth.scala#L201)? `featuresCol` is really a standard trait with `HasFeaturesCol` used in different ml model across Spark.

**Comment by zero323** [comment]
My point is that whenever we use `HasFeaturesCol` it is expected that `featuresCol` is `ml.linalg.VectorUDT`. This creates nice and consistent API. 

If we use it for `FPGrowth` it is no longer so clear. Once / if `PrefixSpan` is implemented we'll get just another input format using the same `trait` / mixin. If you look at SO you'll see that some users  are already confused what is the expected input for ML algorithms and I believe this can make it even worse. Just saying...

**Comment by felixcheung** [comment]
I can see your point, but renaming it only on the R side is not really addressing the issue.
Please feel free to open a JIRA on spark.ml FPGrowth and start a discussion there.

**Comment by zero323** [comment]
You're right of course: [SPARK-19899](https://issues.apache.org/jira/browse/SPARK-19899).

File: R/pkg/R/mllib_fpm.R
Comment (wangmiao1981) [inline R/pkg/R/mllib_fpm.R]: This line seems exceeding the length limit.

File: R/pkg/R/mllib_fpm.R
Comment (wangmiao1981) [inline R/pkg/R/mllib_fpm.R]: Other APIs do not have blank line here. I think we should be consistent.

File: R/pkg/R/mllib_fpm.R
Comment (wangmiao1981) [inline R/pkg/R/mllib_fpm.R]: add blank line. See other examples.

File: R/pkg/R/mllib_fpm.R
Comment (wangmiao1981) [inline R/pkg/R/mllib_fpm.R]: no blank line here.

File: R/pkg/R/mllib_fpm.R
Comment (wangmiao1981) [inline R/pkg/R/mllib_fpm.R]: Add blank line.

File: R/pkg/R/mllib_fpm.R
Comment (wangmiao1981) [inline R/pkg/R/mllib_fpm.R]: Add blank line

File: R/pkg/R/mllib_fpm.R
Comment (wangmiao1981) [inline R/pkg/R/mllib_fpm.R]: add blank line

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: alignment

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: indentation seems incorrect here and above line.

**Comment by felixcheung** [comment]
I suppose we should wait on #17321 before acting on this PR?

**Comment by zero323** [comment]
@felixcheung It is completely up to you. I'll have to patch one or another (and also #17218) and at the end of the day it doesn't make much difference.

**Comment by felixcheung** [comment]
Could you update this PR to have the parameter  itemsCol
And remove predictionCol (if I recall, we don't expose that in the R API for other models either)

**Comment by zero323** [comment]
Done. I marked it as blocked by SPARK-19899 and leave it to be retested once #17321 is merged.

**Comment by zero323** [comment]
Jenkins retest this please.

**Comment by felixcheung** [comment]
this is the error from appveyor build
```
[00:16:14] [ERROR] C:\projects\spark\mllib\src\main\scala\org\apache\spark\ml\r\FPGrowthWrapper.scala:50: value setItemsCol is not a member of org.apache.spark.ml.fpm.FPGrowth
[00:16:14] possible cause: maybe a semicolon is missing before `value setItemsCol'?
[00:16:14] [ERROR]       .setItemsCol(itemsCol)
[00:16:14] [ERROR]        ^
[00:16:29] [ERROR] one error found
```

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: I think we discussed this - let's make it `FP-Growth` or `Frequent Pattern Mining` (https://spark.apache.org/docs/latest/mllib-frequent-pattern-mining.html) as the title

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: can you check if this generate the doc properly
`<\url{http://dx.doi.org/10.1145/1454008.1454027}>`
generally it should be 
`\href{http://...}{Text}`

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: ditto here for url.
In fact, I'm not sure we need to include all the links here but instead link to 
https://spark.apache.org/docs/latest/mllib-frequent-pattern-mining.html

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: we don't generally use this tag. Do you want to move to @seealso, or just link to in the description above

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: `numPartitions` by default is not set in Scala - let's default this to NULL instead here
(but do not as.integer if value is NULL - something like
numPartitions <- if (is.null(numPartitions)) NULL else as.integer(numPartitions)

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: given the earlier suggestion, we should also check numPartition > 0 in R before passing to here

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: anything else we could add as metadata that is not in the model already?

**Review by felixcheung — CHANGES_REQUESTED** [review]
generally good, a few last comments

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: I believe predictionCol param only allow you to change the name of the column - prediction is always still going to be there, no?

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: was https://github.com/apache/spark/pull/17170#discussion_r104736398

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: Sounds good. I'll link the docs.

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: It does render the link as expected, but linking ML docs is indeed a better choice.

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (zero323) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: I don't think so. Model captures all the parameters.

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: Correct if I am wrong but this cannot be done like this. If we want to default to `NULL` (I am not fond of this idea) we have to pass argument as a `character` / `String` and parse it once in JVM.

File: R/pkg/R/mllib_fpm.R
Comment (zero323) [inline R/pkg/R/mllib_fpm.R]: I'll remove it completely and just link to the docs.

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: as this https://github.com/apache/spark/pull/17170/commits/65229163721475f7769387d3e4ba912e570cecc3#r107011745 we should check numPartitions too?
How about changing it to
```
if (!is.null(numPartitions)) {
  numPartitions <- as.integer(numPartitions)
  stopifnot(numPartitions > 0)
}
```

**Review by felixcheung — CHANGES_REQUESTED** [review]

**Comment by felixcheung** [comment]
also would you be updating the R vignettes, ML programming guide and example?

**Comment by zero323** [comment]
Let's make it a separate task. For ML guide we have to wait for #17130 anyway.

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: let's document the list of column like in Python: https://github.com/apache/spark/pull/17218/files#diff-b6dbf16870bd2cca9b4140df8aebd681R121

for reference, see https://github.com/apache/spark/blob/master/R/pkg/R/mllib_clustering.R#L249

File: R/pkg/inst/tests/testthat/test_mllib_fpm.R
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_mllib_fpm.R]: we need to add a test when numPartitions is not set...

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: and this comment https://github.com/apache/spark/pull/17170#discussion_r107011745

Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: and https://github.com/apache/spark/pull/17170#discussion_r107349375

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala
Comment (zero323) [inline mllib/src/main/scala/org/apache/spark/ml/r/FPGrowthWrapper.scala]: If you feel it is necessary. Personally I wanted to treat any non-strictly positive number as `null`.

**Comment by zero323** [comment]
Jenkins retest this please.

**Comment by zero323** [comment]
@felixcheung Looks like some issue with the structured streaming: https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75276/

**Comment by felixcheung** [comment]
Jenkins, retest this please

**Comment by felixcheung** [comment]
@zero323 how about https://github.com/apache/spark/pull/17170#discussion_r108079970

File: R/pkg/R/mllib_fpm.R
Comment (felixcheung) [inline R/pkg/R/mllib_fpm.R]: Actually, sorry we need to change `DataFrame` to `SparkDataFrame` in R

**Review by felixcheung — CHANGES_REQUESTED** [review]

**Review by felixcheung — APPROVED** [review]

**Comment by felixcheung** [comment]
merged to master.
@zero323 could you follow up with vignettes and programming guide update please - we need them for the 2.2.0 release.

**Comment by zero323** [comment]
Of course. Do we have / need a JIRA ticket for that?

**Comment by felixcheung** [comment]
for this, it's optional, but I opened one for tracking purpose https://issues.apache.org/jira/browse/SPARK-20208?filter=12333531

**Pull request closed without merging**
